### PR TITLE
Group 16 - User eviction from cache

### PR DESCRIPTION
**Pull request by Casper-NS**
The cache is now separated by user id.
Changed the cache to be an LRU (Least Recently Used) cache.
This allows users to be evicted once the cache is full. (Default right now is max 100 cached users)

The cache will automatically evict the least recently used user when inserting new users in the cache.

The RwLock has been changed to a mutex, this means only 1 thread can read it at a time. We have tested the benchmarks and the similar performance suggests that this is not an issue.

### CODE REVIEW

File: src/DataReader/component_loader.rs [hunk 1]
```diff
@@ -6,15 +8,26 @@ use crate::ModelObjects::queries::Query;
...
 type ComponentsMap = HashMap<String, Component>;
 
 /// A struct used for caching the models.
-#[derive(Debug, Default, Clone)]
+#[derive(Debug, Clone)]
 pub struct ModelCache {
-    // TODO: A concurrent hashmap may be faster to use and cause less prone to locking, but is not part of the standard library.
-    cache: Arc<RwLock<HashMap<u32, Arc<ComponentsMap>>>>,
+    // TODO: A concurrent lru may be faster to use and cause less prone to lock contention.
+    cache: Arc<Mutex<LruCache<i32, (u32, Arc<ComponentsMap>)>>>,
```
Comment (henneboy) [inline src/DataReader/component_loader.rs:20]: Clippy

**Review by henneboy — APPROVED** [review]
Almost lgtm, plz make clippy happy ;)

**Review by lugialukas — APPROVED** [review]
LGTM

File: src/DataReader/component_loader.rs (hunk 1 above)
Comment (Casper-NS) [reply src/DataReader/component_loader.rs]: Should be fixed now

**Review by t-lohse — DISMISSED** [review]

**Review by henneboy — DISMISSED** [review]
Lgtm

File: src/DataReader/component_loader.rs [hunk 2]
```diff
@@ -7,15 +9,31 @@ use crate::ModelObjects::system_declarations::SystemDeclarations;
...
+    components_map: Arc<ComponentsMap>,
+}
+
 /// A struct used for caching the models.
-#[derive(Debug, Default, Clone)]
+#[derive(Debug, Clone)]
 pub struct ModelCache {
-    // TODO: A concurrent hashmap may be faster to use and cause less prone to locking, but is not part of the standard library.
-    cache: Arc<RwLock<HashMap<u32, Arc<ComponentsMap>>>>,
+    // TODO: A concurrent lru may be faster to use and cause less prone to lock contention.
```
Comment (S3JER) [inline src/DataReader/component_loader.rs:25]: Noget i vil tilføje nu, @t-lohse plejer ikke være fan af at merge todoes ind i main

**Review by S3JER — APPROVED** [review]

File: src/DataReader/component_loader.rs (hunk 2 above)
Comment (mads256h) [reply src/DataReader/component_loader.rs:25]: Nej det var mere for at sige at det er måske muligt at improve performance endnu mere

**Review by WassawRoki — APPROVED** [review]

**Pull request merged**